File: Olympiad/COCI/official/2008/contest3_solutions/redoks.cpp
#include <cstdio>
#include <cstring>

#define MAX 250000

inline int mod10( int x ) { 
   if( x < 10 ) return x;
   return x-10;
}

struct Interval {
   int lo, hi;
   int freq[10];   
   int lazy;
   Interval *left, *right;
   
   Interval() { lazy = 0; memset( freq, 0, sizeof freq ); }

   void rotate( int rot ) {
      static int tmp[10];
      for( int i = 0; i < 10; ++i ) tmp[i] = freq[i];
      for( int i = 0; i < 10; ++i ) freq[mod10(i+rot)] = tmp[i];
      lazy = mod10( lazy + rot );
   }
   
   void split() {
      left->rotate( lazy );
      right->rotate( lazy );
      lazy = 0;
   }
   
   void join() {
      for( int i = 0; i < 10; ++i ) 
         freq[i] = left->freq[i] + right->freq[i];
   }
} memory[2*MAX], *allocator = memory;

int n;
char start[MAX+1];

void tree_build( Interval *x ) {
   if( x->lo == x->hi ) {
      x->freq[ start[x->lo]-'0' ] = 1;
   } else {
      int mid = (x->lo + x->hi)/2;
      
      x->left = allocator++;
      x->left->lo = x->lo; 
      x->left->hi = mid;
      tree_build( x->left );

      x->right = allocator++;
      x->right->lo = mid+1; 
      x->right->hi = x->hi;
      tree_build( x->right );

      x->join();
   }
}

int A, B;
int ret;

void tree_query( Interval *x ) {
   if( x->lo > B || x->hi < A ) return;
   if( A <= x->lo && x->hi <= B ) {
      for( int i = 0; i < 10; ++i ) ret += i * x->freq[i];
      x->rotate( 1 );
   } else {
      x->split();
      tree_query( x->left );
      tree_query( x->right );
      x->join();
   }
}

int main( void ) {
   int n, m;

   scanf( "%d%d ", &n, &m );
   gets( start );

   Interval *root = allocator++;
   root->lo = 0;
   root->hi = n-1;

   tree_build( root );
   
   for( int i = 0; i < m; ++i ) {
      scanf( "%d%d", &A, &B ); --A; --B;
      
      ret = 0;
      tree_query( root );      
      printf( "%d\n", ret );
   }

   return 0;
}

File: Olympiad/COCI/official/2008/contest3_solutions/redoks.pas
program redoks;

const MAX = 250000;

function mod10( x : longint ) : longint;
begin
   if x < 10 then mod10 := x else mod10 := x-10;
end;

type
   PInterval = ^Interval;
   Interval = record
      lo, hi : longint;
      freq : array[0..9] of longint;
      lazy : longint;
      left, right : PInterval;
   end;

var tmp : array [0..9] of longint;

procedure rotate( x : PInterval; rot : longint );
var i : longint;
begin
   for i := 0 to 9 do tmp[i] := x^.freq[i];
   for i := 0 to 9 do x^.freq[mod10(i+rot)] := tmp[i];
   x^.lazy := mod10( x^.lazy + rot );
end;

procedure split( x : PInterval );
begin
   rotate( x^.left, x^.lazy );
   rotate( x^.right, x^.lazy );
   x^.lazy := 0;
end;

procedure join( x : PInterval );
var i : longint;
begin
   for i := 0 to 9 do
      x^.freq[i] := x^.left^.freq[i] + x^.right^.freq[i];
   x^.lazy := 0;
end;

var
   i, n, m : longint;
   start : array [1..MAX] of char;
   root, allocator : Pinterval;
   memory : array [0..2*MAX] of Interval;
   ret : longint;
   A, B : longint;


procedure treeBuild( x : PInterval );
var i, mid : longint;
begin
   if x^.lo = x^.hi then begin

      for i := 0 to 9 do x^.freq[i] := 0;

      x^.freq[ ord(start[x^.lo]) - 48 ] := 1;
      x^.lazy := 0;

   end else begin

      mid := (x^.lo + x^.hi) div 2;

      x^.left := allocator; inc( allocator );
      x^.left^.lo := x^.lo;
      x^.left^.hi := mid;
      treeBuild( x^.left );

      x^.right := allocator; inc( allocator );
      x^.right^.lo := mid+1;
      x^.right^.hi := x^.hi;
      treeBuild( x^.right );

      join( x );

   end;
end;

procedure treeQuery( x : PInterval );
var i : longint;
begin
   if (x^.lo > B) or (x^.hi < A) then exit;
   if (A <= x^.lo) and (x^.hi <= B) then begin

      for i := 0 to 9 do ret := ret + i * x^.freq[i];
      rotate( x, 1 );

   end else begin

      split( x );

      treeQuery( x^.left );
      treeQuery( x^.right );

      join( x );

   end;
end;

begin
   readln( n, m );
   for i := 1 to n do
      read( start[i] );
   readln;

   allocator := @memory[0];
   root := allocator; inc( allocator );

   root^.lo := 1;
   root^.hi := n;
   treeBuild( root );

   for i := 1 to m do begin
      readln( A, B );

      ret := 0;
      treeQuery( root );
      writeln( ret );
   end;
end.
